cheatsheets/gleam-for-go-users.md

```diff
@@ -0,0 +1,1460 @@
+### Records
+
+Custom type allows you to define a collection data type with a fixed number of
+named fields, and the values in those fields can be of differing types.
+
+#### Go
+
+Go uses structs to define user-defined, record-like types.
+Properties are defined as fields and initial values are generally set in
+the literal form. Any unset fields during construction are defaulted to
+their zero value.
+
+Go does not use constructors, but many Go developers will create one or more
+functions to construct a struct in an ergonomic way.
+
+```Go
+type Person struct {
+  name string
+  age int
+}
+
+func NewPerson(name string, age int) Person {
+    return Person{ name, age } // Short hand literal construction
+    // Could also be done with
+    // return Person { name: name, age: age }
+}
+
+// Inside of a function
+person := NewPerson("Joe", 40);
+person.name // Joe;
+```
+
+#### Gleam
+
+Gleam's custom types can be used as structs. At runtime, they have a tuple
+representation and are compatible with Erlang records (or JavaScript objects).
+
+```gleam
+type Person {
+  Person(name: String, age: Int)
+}
+
+let person = Person(name: "Joe", age: 40)
+let name = person.name
+```
+
+An important difference to note is there is no Java-style object-orientation in
+Gleam, thus methods can not be added to types. However opaque types exist,
+see below.
+
+### Unions
+
+Go generally does not support unions, with a small exception to that.
+
+In Gleam functions must always take and receive one type. To have a union of
+two different types they must be wrapped in a new custom type.
+
+#### Go
+
+Generally the usecase of wanting to use multiple concrete types is solved with
+Interfaces.
+
+```Go
+type Stringer interface {
+	String() string // Types must have a function with this signature defined
+}
+
+type StructA struct {
+	name string
+	age int
+}
+
+func (s StructA) String() string {  // This method fulfills the Stringer interface
+	return fmt.Sprintf("%s,%d", s.name, s.age)
+}
+
+type StructB struct {
+	x int
+	y int
+}
+
+func (s StructB) String() string {  // This method fulfills the Stringer interface
+	return fmt.Sprintf("%d,%d", s.x, s.y)
+}
+
+func print(str Stringer) {
+	fmt.Print(str.String())
+}
+
+func main() {
+	print(StructA{ "Bob", 47}) // Prints Bob,47
+	print(StructB{ 42,0 }) // Prints 42,0
+}
+```
+
+The only place where Go uses something related to true unions or sum types is in Type Sets,
+which are constraints on a Generic.
+
+```go
+type Num interface {
+	int | float
+}
+```
+
+Unfortunately, these unions provide little value to the Go Developer due to their
+constraints.
+
+#### Gleam
+
+```gleam
+type IntOrFloat {
+  AnInt(Int)
+  AFloat(Float)
+}
+
+fn int_or_float(X) {
+  case X {
+    True -> AnInt(1)
+    False -> AFloat(1.0)
+  }
+}
+```
+
+### Opaque custom types
+
+In Go, you can choose to not export a type and then they will be considered
+opaque to the user of your package.  Just have the type have a lowercase
+identifier.
```

Comment:
Can you still reference the type from other packages? I thought it would make it private, not opaque

Reply:
It will be opaque from outside the package, which is the same thing as private? I might be misunderstanding

Reply:
Sorry, I should clear this up.  To make a type Opaque in Go is having the type public, but all fields and methods being private.  It would then be unusable outside of whatever public functions that work with it are exposed.  The water is muddied by the fact that methods are just functions that have a certain type as the receiver.  But I think not exposing them is closer to the intent of opaque types in that they may flow through code outside the package, but that code should have no idea about the details of the type.